src/shell/xdg/mod.rs

```diff
@@ -1,58 +1,142 @@
 //! ## Cross desktop group (XDG) shell
 // TODO: Examples
 
-use wayland_client::{ConnectionHandle, QueueHandle};
+use std::marker::PhantomData;
+
+use wayland_backend::client::InvalidId;
+use wayland_client::{protocol::wl_surface, ConnectionHandle, Dispatch, QueueHandle};
 use wayland_protocols::xdg_shell::client::{xdg_surface, xdg_wm_base};
 
 mod inner;
 pub mod popup;
 pub mod window;
 
+#[derive(Debug, thiserror::Error)]
+pub enum XdgSurfaceError {
+    /// The xdg_wm_base global is not available.
+    #[error("the xdg_wm_base global is not available")]
+    MissingRequiredGlobals,
+
+    /// Protocol error.
+    #[error(transparent)]
+    Protocol(#[from] InvalidId),
+}
+
 #[derive(Debug)]
-pub struct XdgShellState {
+pub struct XdgShellState<D> {
```

Comment:
Why is this type parameter needed? Wouldn't making `create_xdg_surface` generic instead work?

Reply:
```
error[E0207]: the type parameter `D` is not constrained by the impl trait, self type, or predicates
  --> src/shell/xdg/inner.rs:72:6
   |
72 | impl<D> DelegateDispatchBase<xdg_surface::XdgSurface> for XdgShellState
   |      ^ unconstrained type parameter
```

for
```rust
impl<D> DelegateDispatchBase<xdg_surface::XdgSurface> for XdgShellState {
    type UserData = XdgSurfaceData<D>;
}
```

Reply:
Oh, right :+1: